api: support getting repository Git tree (#5934)

bharatnc opened the pull request:
This PR adds the tree  api endpoint to gogs api:
`GET/repos/:owner/:repo/git/trees/:tree_sha`

This new api endpoint that is being added is in conformance to
the GitHub REST API v3 specification. Documentation can be found
here: developer.github.com/v3/git/trees/#get-a-tree

For a given user, repo and sha value, this api (currently) returns
a single tree using the SHA1 value for that tree.

- Recursive implementation is yet to be implemented.
- Creating a Tree using POST is yet to be implemented.

Example curl:

```

l curl -H "Authorization: token REDACTED" http://localhost:3000/api/v1/repos/root/testrepo/git/trees/c59441ded1549b149def0d4c54594d31a7f3718f -X GET  | jq .
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   940  100   940    0     0  12034      0 --:--:-- --:--:-- --:--:-- 12051

[
  {
    "sha": "c59441ded1549b149def0d4c54594d31a7f3718f",
    "tree": [
      {
        "mode": "120000",
        "path": "/home/bharatnc/gogs-repositories/root/testrepo.git",
        "sha": "472ac2361b65136b393d652de25341e2ea44f299",
        "size": 1077,
        "type": "blob",
        "url": "http://localhost:3000/api/v1/repos/root/testrepo/git/trees/472ac2361b65136b393d652de25341e2ea44f299"
      },
      {
        "mode": "120000",
        "path": "/home/bharatnc/gogs-repositories/root/testrepo.git",
        "sha": "70fcb456d436f08462602f26df6fb7e167e7a916",
        "size": 12,
        "type": "blob",
        "url": "http://localhost:3000/api/v1/repos/root/testrepo/git/trees/70fcb456d436f08462602f26df6fb7e167e7a916"
      },
      {
        "mode": "120000",
        "path": "/home/bharatnc/gogs-repositories/root/testrepo.git",
        "sha": "092c58d4b63df5779a4d020b1fdbb762421bbb4f",
        "size": 380,
        "type": "blob",
        "url": "http://localhost:3000/api/v1/repos/root/testrepo/git/trees/092c58d4b63df5779a4d020b1fdbb762421bbb4f"
      }
    ],
    "url": "http://localhost:3000/api/v1/repos/root/testrepo/git/trees/c59441ded1549b149def0d4c54594d31a7f3718f"
  }
]

```

bharatnc commented:
@unknwon , I made some progress with respect to the trees api endpoint for gogs based on your inputs. Would appreciate feedback. Thank you!

unknwon commented on internal/route/api/v1/repo/file.go:
```diff
@@ -45,6 +48,72 @@ func GetArchive(c *context.APIContext) {
+		c.ServerError("OpenRepository", err)
+		return
+	}
+	c.Repo.GitRepo = gitRepo
```
`c.Repo.GitRepo` is already set in `context.RepoRef()` call, so you can just use it.

bharatnc replied:
Thanks, fixed it.

unknwon commented on internal/route/api/v1/repo/file.go:
```diff
@@ -45,6 +48,72 @@ func GetArchive(c *context.APIContext) {
+		return
+	}
+	if entries == nil {
+		c.NotFound()
```
Is it really not found? Or it's just an empty tree/directory.

bharatnc replied:
Yea, you're right. It's shouldn't be not found, it should be just an empty tree. I am going to preserve the top level structure but return an empty tree. 

Which will ook like:
```bash
[
  {
    "url":"http://localhost:3000/api/v1/repos/root/testrepo/git/trees/c59441ded1549b149def0d4c54594d31a7f3718f"
    "sha": "c59441ded1549b149def0d4c54594d31a7f3718f",
    "tree": []
}
```

bharatnc commented:
@unknwon,

> I don't see you add the new import of golang.org/x/tools and githubo.com/modern-go/* anywhere in the diffs, why go.mod wants to include those?
They should be gone now. Not sure what happened there.

I don't see any changes to "public/", "templates/" directories and any "*.less" files, where does these changes come form?
> Yep, interesting, I have discarded changes to those files and updated the PR. In fact when I again ran `make web`, I could see that new files are generated - though there are no changes to the UI :( 

```bash➜  gogs git:(bnc/gogs-5934-git-tree-api) ✗ make web
rm -f internal/assets/public/public_gen.go
go generate internal/assets/public/public.go
gofmt -s -w internal/assets/public/public_gen.go
```

> Which IDE you're using? The "internal/route/api/v1/repo/file.go" looks like is not gofmted, and it would more ideal to put in its own file, namely "tree.go".

i use GoLand by Jetbrains. But that shouldn't matter, I usually run `gofmt -w <file>` to ensure formatting. Only in this file it did not make any difference. I have made a few changes and formatted the file again so that it looks normal now.

Thank you for the feedback. I have also moved this new Tree method into it's own file called  `tree.go` as per your suggestion. Please let me know if you would like me to do some other changes. Thank you!

unknwon commented:
> > I don't see you add the new import of golang.org/x/tools and githubo.com/modern-go/* anywhere in the diffs, why go.mod wants to include those?
> > They should be gone now. Not sure what happened there.
> 
> I don't see any changes to "public/", "templates/" directories and any "*.less" files, where does these changes come form?

I think it was my mistake that I didn't properly update these files on `master`, which I fixed yesterday, so no longer showing them on your PR.

unknwon commented:
> i use GoLand by Jetbrains. But that shouldn't matter, I usually run `gofmt -w <file>` to ensure formatting. Only in this file it did not make any difference. I have made a few changes and formatted the file again so that it looks normal now.

GoLand sadly do not enable `gofmt` by default, here's how you could tell it to format:
![image](https://user-images.githubusercontent.com/2946214/75218895-580baf00-57d6-11ea-9d01-96600b075113.png)

bharatnc commented:
I have also synced my fork with the master and have properly formatted the changes.

bharatnc commented:
> > i use GoLand by Jetbrains. But that shouldn't matter, I usually run `gofmt -w <file>` to ensure formatting. Only in this file it did not make any difference. I have made a few changes and formatted the file again so that it looks normal now.
> 
> GoLand sadly do not enable `gofmt` by default, here's how you could tell it to format:
> ![image](https://user-images.githubusercontent.com/2946214/75218895-580baf00-57d6-11ea-9d01-96600b075113.png)

Oh yea, I usually manually do `Tools ->  Go Tools -> Go fmt file`

unknwon commented on internal/route/api/v1/repo/tree.go:
```diff
@@ -0,0 +1,65 @@
+	"gogs.io/gogs/internal/db"
+)
+
+func GetRepoTree(c *context.APIContext) {
```
```suggestion
func GetRepoGitTree(c *context.APIContext) {
```

bharatnc replied:
Sure, changed.

unknwon commented on internal/route/api/v1/repo/tree.go:
```diff
@@ -0,0 +1,65 @@
+		c.ServerError("GetRepoTree", err)
+		return
+	}
+	var results []map[string]interface{}
```
Use `[]map[string]interface{}` in this case is not ideal, should define concrete types:

```go
type repoGitTreeResponse struct {
	Sha  string `json:"sha"`
	URL  string `json:"url"`
	Tree []*repoGitTree `json:"tree"`
}

type repoGitTree struct{
	Path string `json:"path"`
	Mode string `json:"mode"`
	Type string `json:"type"`
	Size int    `json:"size,omitempty"`
	Sha  string `json:"sha"`
	URL  string `json:"url"`
}
```

_Generated using https://mholt.github.io/json-to-go/_

bharatnc replied:
I agree, these should be now implemented, verified that these still return the same response:
```bash
{
  "sha": "c59441ded1549b149def0d4c54594d31a7f3718f",
  "url": "http://localhost:3000/api/v1/repos/root/testrepo/git/trees/c59441ded1549b149def0d4c54594d31a7f3718f",
  "tree": [
    {
      "path": "/home/bharatnc/gogs-repositories/root/testrepo.git",
      "mode": "120000",
      "type": "blob",
      "size": 1077,
      "sha": "472ac2361b65136b393d652de25341e2ea44f299",
      "url": "http://localhost:3000/api/v1/repos/root/testrepo/git/trees/472ac2361b65136b393d652de25341e2ea44f299"
    },
    {
      "path": "/home/bharatnc/gogs-repositories/root/testrepo.git",
      "mode": "120000",
      "type": "blob",
      "size": 12,
      "sha": "70fcb456d436f08462602f26df6fb7e167e7a916",
      "url": "http://localhost:3000/api/v1/repos/root/testrepo/git/trees/70fcb456d436f08462602f26df6fb7e167e7a916"
    },
    {
      "path": "/home/bharatnc/gogs-repositories/root/testrepo.git",
      "mode": "120000",
      "type": "blob",
      "size": 380,
      "sha": "092c58d4b63df5779a4d020b1fdbb762421bbb4f",
      "url": "http://localhost:3000/api/v1/repos/root/testrepo/git/trees/092c58d4b63df5779a4d020b1fdbb762421bbb4f"
    }
  ]
}

```

unknwon commented:
PS: you don’t need to force push, we’re going to squash and merge anyway.

unknwon commented on internal/route/api/v1/repo/tree.go:
```diff
@@ -0,0 +1,79 @@
+	repoPath := db.RepoPath(c.Params(":username"), c.Params(":reponame"))
+	gitTree, err := c.Repo.GitRepo.GetTree(c.Params(":sha"))
+	if err != nil {
+		c.NotFoundOrServerError("GetRepoTree", git.IsErrNotExist, err)
```
```suggestion
		c.NotFoundOrServerError("GetRepoGitTree", git.IsErrNotExist, err)
```

unknwon commented on internal/route/api/v1/repo/tree.go:
```diff
@@ -0,0 +1,79 @@
+	}
+	entries, err := gitTree.ListEntries()
+	if err != nil {
+		c.ServerError("GetRepoTree", err)
```
```suggestion
		c.ServerError("GetRepoGitTree", err)
```

bharatnc commented:
> PS: you don’t need/have to force push, we’re going to squash and merge anyway.

Noted - will maintain additional commits hereafter. I'm just used to maintaining a single commit (force pushing) on a daily basis :) .

unknwon commented on internal/route/api/v1/repo/tree.go:
```diff
@@ -0,0 +1,79 @@
+	var children []*repoGitTree
+	var mode string
+	templateURL := `%s/repos/%s/%s/git/trees`
+	templateURL = fmt.Sprintf(templateURL, c.BaseURL, c.Params(":username"), c.Params(":reponame"))
```
Why not just 

```go
templateURL := fmt.Sprintf("%s/repos/%s/%s/git/trees", c.BaseURL, c.Params(":username"), c.Params(":reponame"))
```

unknwon commented on internal/route/api/v1/repo/tree.go:
```diff
@@ -0,0 +1,79 @@
+			Sha: c.Params(":sha"),
+			URL: fmt.Sprintf(templateURL+"/%s", c.Params(":sha")),
+		}
+		c.JSONSuccess(res)
```
You need a `return` after this.

unknwon replied:
We could go to the extreme:
```go
c.JSONSuccess(&repoGitTreeResponse{
	Sha: c.Params(":sha"),
	URL: fmt.Sprintf(templateURL+"/%s", c.Params(":sha")),
})
return
```

unknwon commented on internal/route/api/v1/repo/tree.go:
```diff
@@ -0,0 +1,78 @@
+		return
+	}
+	var children []*repoGitTree
+	var mode string
```
We don't use these two variables until after the `if entries == nil {`, move them down to the their use cases is better.

unknwon commented on internal/route/api/v1/repo/tree.go:
```diff
@@ -0,0 +1,78 @@
+		}
+		c.JSONSuccess(res)
+	}
+	for _, entry := range entries {
```
An empty line before the for loop is better for visually separating the logic pieces.

unknwon commented on internal/route/api/v1/repo/tree.go:
```diff
@@ -0,0 +1,78 @@
+	"gogs.io/gogs/internal/db"
+)
+
+type repoGitTreeResponse struct {
```
Sorry I just came up a better name for it 😝 
```suggestion
type repoGitTree struct {
```

unknwon commented on internal/route/api/v1/repo/tree.go:
```diff
@@ -0,0 +1,78 @@
+	Tree []*repoGitTree `json:"tree"`
+}
+
+type repoGitTree struct {
```
```suggestion
type repoGitTreeEntry struct {
```

unknwon commented on internal/route/api/v1/repo/tree.go:
```diff
@@ -0,0 +1,78 @@
+		URL:  fmt.Sprintf(templateURL+"/%s", c.Params(":sha")),
+		Tree: children,
+	}
+	c.JSONSuccess(results)
```
Again, we can go to the extreme:

```go
c.JSONSuccess( &repoGitTreeResponse{
		Sha:  c.Params(":sha"),
		URL:  fmt.Sprintf(templateURL+"/%s", c.Params(":sha")),
		Tree: children,
	})
```

unknwon commented on internal/route/api/v1/repo/tree.go:
```diff
@@ -0,0 +1,79 @@
+			mode = ""
+		}
+		children = append(children, &repoGitTreeEntry{
+			Path: repoPath,
```
Since we don't support recursive yet, I believe the `Path` is simply `entry.Name()`.

Exposing server path is actually a security problem :)

unknwon commented:
FYI, I directly edited the file for some minor changes. Please see my comment above, that should be the last thing we need to address!

bharatnc commented:
> FYI, I directly edited the file for some minor changes. Please see my comment above, that should be the last thing we need to address!

@unknwon  thanks and I have addressed the last bit of the change that's required. I really appreciate the reviews!

unknwon commented:
Congrats, merging!

bharatnc commented:
@unknwon I can work on adding the docs for this new endpoint.

unknwon commented:
Great :)
